### PR TITLE
[EH] Add --experimental-new-eh option to wasm-opt

### PR DESCRIPTION
**Pull request by aheejin**
This adds `--experimental-new-eh` option to `wasm-opt`. The difference between this and `--translate-to-new-eh` is, `--translate-to-new-eh` just runs `TranslateToNewEH` pass, while `--experimental-new-eh` attaches `TranslateToNewEH` pass at the end of the whole optimization pipeline. So if no other passes or optimization options (`-On`) are specified, it is equivalent to `--translate-to-new-eh`. If other optimization passes are specified, it runs them and at the end run the translator to ensure the new EH instructions are emitted. The reason we are doing this this way is that the optimization pipeline as a whole does not support the new EH instruction yet, but we would like to provide an option to emit a reasonably OK code with the new EH instructions.

This also means when the optimization level > 3, it will also run the StackIR + local2stack optimization after the translation.

Not sure how to test the output of this option, given that there is not much point in testing the default optimization passes, and it is also not clear how to print the stack IR if the stack ir generation and optimization runs as a part of the pipeline and not the explicit command line options.

This is created in favor of #6267, which added the option to `optimization-options.h`. It had a problem of running the translator multiple times when `-On` was given multiple times in the command line, which I learned was rather a common usage. This adds the option directly to `wasm-opt.cpp`, which avoids the problem. With this, it is still possible to create and optimize Stack IR unnecessarily, but that feels a better alternative.

### CODE REVIEW

File: src/tools/wasm-opt.cpp [hunk 1]
```diff
@@ -398,6 +413,26 @@ int main(int argc, const char* argv[]) {
     }
   }
 
+  if (translateToNewEH) {
+    BYN_TRACE("translating to new EH instructions...\n");
+    PassRunner runner(&wasm, options.passOptions);
+    runner.add("translate-to-new-eh");
+    // perform Stack IR optimizations here, at the very end of the
+    // optimization pipeline
```
Comment (kripken) [inline src/tools/wasm-opt.cpp:421]:
```suggestion
    // Perform Stack IR optimizations here, at the very end of the
    // optimization pipeline.
```

**Review by kripken — COMMENTED** [review]
Code lgtm.

I think a test would be good to add just to be safe. Maybe we can check in the result of `--experiment-new-eh -O` on a binary, and a lit test could just check that the result has not changed? (just a binary diff with a boolean outcome)

**Comment by aheejin** [comment]
> Code lgtm.
> 
> I think a test would be good to add just to be safe. Maybe we can check in the result of `--experiment-new-eh -O` on a binary, and a lit test could just check that the result has not changed? (just a binary diff with a boolean outcome)

Not sure if I understand. What binary? (Old EH instructions or new ones?) Why does it have to be unchanged? And why would it be a binary test rather than a text one?

I can add a test showing that `--experimental-new-eh` does the same thing (translation) w/ `--translate-to-new-eh` when given alone. It is hard to test the combination of the translation and stack IR optimization because there is no way I can print the stack IR result with this.

**Comment by kripken** [comment]
I mean taking a file `old.wasm` with old wasm EH, running `wasm-opt --experimental-new-eh -O -o new.wasm`, and checking in both `old.wasm` and `new.wasm`. The test would run that same command and see that it is the expected binary.

If we had wabt available then we could diff the text of the output, but without that, I think we can just compare the binaries as a boolean. That is, in this PR we can manually verify with wabt or another tool that the binary is properly optimized, including stackIR opts have run. Then the test will make sure we don't regress on that test.

**Review by kripken — APPROVED** [review]

**Review by tlively — COMMENTED** [review]
Instead of checking in binaries for the tests, could we have the lit test assemble input.wat using both `--experimental-new-eh` and a manual pipeline using `--translate-to-new-eh` and check that the results are the same?

Other than that, LGTM.

**Comment by aheejin** [comment]
> Instead of checking in binaries for the tests, could we have the lit test assemble input.wat using both `--experimental-new-eh` and a manual pipeline using `--translate-to-new-eh` and check that the results are the same?

Do you mean
```console
wasm-opt --experimental-new-eh [INPUT] -o [OUTPUT]
```
and
```console
wasm-opt --translate-to-new-eh [INPUT] -o [OUTPUT]
```
produce the same output binary (from the given input wat file), when each of these is given as a standalone flag?

**Comment by tlively** [comment]
Yes, exactly. And maybe you could add the -O1 and --generate-stack-ir --optimize-stack-ir flags to get them to run the same optimization pipeline as well.

**Comment by aheejin** [comment]
@tlively PTAL 😀

File: test/lit/passes/experimental-new_eh.wast [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+;; RUN: wasm-opt %s -all --translate-to-new-eh -S -o %a.wasm
+;; RUN: wasm-opt %s -all --experimental-new-eh -S -o %b.wasm
+;; RUN: diff %a.wasm %b.wasm
+
+;; When given with other flags, --experimental-new-eh runs the translator after
+;; running other passes. If --optimize-level >=3, --experimenal-new-eh also runs
+;; StackIR (+ local2stack) optimization. So running '-O --experimental-new-eh'
+;; should be the same as running all these passes separately.
+;; RUN: wasm-opt %s -all -O --translate-to-new-eh -o %t.wasm
+;; RUN: wasm-opt %t.wasm -all --optimize-level=3 --generate-stack-ir --optimize-stack-ir -o %a.wasm
```
Comment (tlively) [inline test/lit/passes/experimental-new_eh.wast:11]: Can we do without the intermediate file here? (@kripken, I don't quite remember how the optimization-level and -O flags interact these days)

**Review by tlively — APPROVED** [review]
LGTM with that comment, unless we can't get rid of the intermediate file, in which case LGTM as-is.

File: test/lit/passes/experimental-new_eh.wast (hunk 2 above)
Comment (aheejin) [reply test/lit/passes/experimental-new_eh.wast:11]: I think we can do without it... Not sure why I thought otherwise. Will fix.

Comment (aheejin) [reply test/lit/passes/experimental-new_eh.wast]: Removed the intermediate file. Will merge this. Thanks!

**Pull request merged**